File: ios/Classes/LiveLocationTrackerPlusPlugin.swift
import Flutter
import UIKit
import CoreLocation

/// Main Flutter plugin class for live_location_tracker_plus on iOS.
///
/// Handles method channel routing, event channel streams, and delegates
/// to specialized managers for location, geofencing, Firebase sync, and permissions.
public class LiveLocationTrackerPlusPlugin: NSObject, FlutterPlugin {

    private var channel: FlutterMethodChannel?
    private var locationEventChannel: FlutterEventChannel?
    private var geofenceEventChannel: FlutterEventChannel?

    private var locationManager: LLTLocationManager?
    private var geofenceManager: LLTGeofenceManager?
    private var permissionHandler: LLTPermissionHandler?
    private var firebaseSyncManager: LLTFirebaseSyncManager?

    private let locationStreamHandler = LocationStreamHandler()
    private let geofenceStreamHandler = GeofenceStreamHandler()

    private var isTrackingActive = false

    override init() {
        channel = nil
        locationEventChannel = nil
        geofenceEventChannel = nil
        locationManager = nil
        geofenceManager = nil
        permissionHandler = nil
        firebaseSyncManager = nil
        super.init()
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = LiveLocationTrackerPlusPlugin()

        // Method channel
        let channel = FlutterMethodChannel(
            name: "live_location_tracker_plus",
            binaryMessenger: registrar.messenger()
        )
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)

        // Location event channel
        let locationEC = FlutterEventChannel(
            name: "live_location_tracker_plus/location_stream",
            binaryMessenger: registrar.messenger()
        )
        locationEC.setStreamHandler(instance.locationStreamHandler)
        instance.locationEventChannel = locationEC

        // Geofence event channel
        let geofenceEC = FlutterEventChannel(
            name: "live_location_tracker_plus/geofence_stream",
            binaryMessenger: registrar.messenger()
        )
        geofenceEC.setStreamHandler(instance.geofenceStreamHandler)
        instance.geofenceEventChannel = geofenceEC

        // Initialize managers
        instance.firebaseSyncManager = LLTFirebaseSyncManager()
        instance.locationManager = LLTLocationManager(
            streamHandler: instance.locationStreamHandler,
            firebaseSyncManager: instance.firebaseSyncManager
        )
        instance.geofenceManager = LLTGeofenceManager(
            streamHandler: instance.geofenceStreamHandler
        )
        instance.permissionHandler = LLTPermissionHandler()
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)

        case "startTracking":
            handleStartTracking(call: call, result: result)

        case "stopTracking":
            handleStopTracking(result: result)

        case "getCurrentLocation":
            handleGetCurrentLocation(call: call, result: result)

        case "isTracking":
            result(isTrackingActive)

        case "addGeofence":
            handleAddGeofence(call: call, result: result)

        case "removeGeofence":
            handleRemoveGeofence(call: call, result: result)

        case "getActiveGeofences":
            handleGetActiveGeofences(result: result)

        case "requestPermission":
            handleRequestPermission(result: result)

        case "checkPermission":
            handleCheckPermission(result: result)

        case "requestBackgroundPermission":
            handleRequestBackgroundPermission(result: result)

        case "enableFirebaseSync":
            handleEnableFirebaseSync(call: call, result: result)

        case "disableFirebaseSync":
            handleDisableFirebaseSync(result: result)

        case "setTrackingMode":
            handleSetTrackingMode(call: call, result: result)

        case "openLocationSettings":
            handleOpenLocationSettings(result: result)

        case "openAppSettings":
            handleOpenAppSettings(result: result)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Tracking

    private func handleStartTracking(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(false)
            return
        }

        let intervalMs = args["intervalMs"] as? Int ?? 5000
        let distanceFilter = args["distanceFilter"] as? Double ?? 10.0
        let accuracy = args["accuracy"] as? Int ?? 2
        let trackingMode = args["trackingMode"] as? Int ?? 1
        let enableAutoPause = args["enableAutoPause"] as? Bool ?? true

        locationManager?.startTracking(
            intervalMs: intervalMs,
            distanceFilter: distanceFilter,
            accuracy: accuracy,
            trackingMode: trackingMode,
            enableAutoPause: enableAutoPause
        )

        isTrackingActive = true
        result(true)
    }

    private func handleStopTracking(result: @escaping FlutterResult) {
        locationManager?.stopTracking()
        isTrackingActive = false
        result(true)
    }

    private func handleGetCurrentLocation(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]
        let accuracy = args?["accuracy"] as? Int ?? 2

        locationManager?.getCurrentLocation(accuracy: accuracy) { locationMap in
            if let locationMap = locationMap {
                result(locationMap)
            } else {
                result(FlutterError(code: "LOCATION_ERROR",
                                   message: "Failed to get current location",
                                   details: nil))
            }
        }
    }

    private func handleSetTrackingMode(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let mode = args["mode"] as? Int else {
            result(false)
            return
        }
        locationManager?.setTrackingMode(mode)
        result(true)
    }

    // MARK: - Geofencing

    private func handleAddGeofence(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let id = args["id"] as? String,
              let latitude = args["latitude"] as? Double,
              let longitude = args["longitude"] as? Double,
              let radius = args["radius"] as? Double else {
            result(false)
            return
        }

        let triggers = args["triggers"] as? [Int] ?? [0, 1]

        let success = geofenceManager?.addGeofence(
            id: id,
            latitude: latitude,
            longitude: longitude,
            radius: radius,
            triggers: triggers
        ) ?? false

        result(success)
    }

    private func handleRemoveGeofence(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let id = args["id"] as? String else {
            result(false)
            return
        }
        let success = geofenceManager?.removeGeofence(id: id) ?? false
        result(success)
    }

    private func handleGetActiveGeofences(result: @escaping FlutterResult) {
        let geofences = geofenceManager?.getActiveGeofences() ?? []
        result(geofences)
    }

    // MARK: - Permissions

    private func handleRequestPermission(result: @escaping FlutterResult) {
        permissionHandler?.requestWhenInUsePermission { status in
            result(status)
        }
    }

    private func handleCheckPermission(result: @escaping FlutterResult) {
        let status = permissionHandler?.checkPermission() ?? 0
        result(status)
    }

    private func handleRequestBackgroundPermission(result: @escaping FlutterResult) {
        permissionHandler?.requestAlwaysPermission { status in
            result(status)
        }
    }

    // MARK: - Firebase Sync

    private func handleEnableFirebaseSync(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let userId = args["userId"] as? String else {
            result(false)
            return
        }

        let collectionPath = args["collectionPath"] as? String ?? "live_locations"
        let syncIntervalMs = args["syncIntervalMs"] as? Int ?? 10000

        let success = firebaseSyncManager?.enable(
            collectionPath: collectionPath,
            userId: userId,
            syncIntervalMs: syncIntervalMs
        ) ?? false

        result(success)
    }

    private func handleDisableFirebaseSync(result: @escaping FlutterResult) {
        firebaseSyncManager?.disable()
        result(true)
    }

    // MARK: - Settings

    private func handleOpenLocationSettings(result: @escaping FlutterResult) {
        if let url = URL(string: UIApplication.openSettingsURLString) {
            DispatchQueue.main.async {
                UIApplication.shared.open(url, options: [:]) { success in
                    result(success)
                }
            }
        } else {
            result(false)
        }
    }

    private func handleOpenAppSettings(result: @escaping FlutterResult) {
        if let url = URL(string: UIApplication.openSettingsURLString) {
            DispatchQueue.main.async {
                UIApplication.shared.open(url, options: [:]) { success in
                    result(success)
                }
            }
        } else {
            result(false)
        }
    }
}

// MARK: - Stream Handlers

/// Stream handler for location updates sent to Flutter.
class LocationStreamHandler: NSObject, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?

    override init() {
        eventSink = nil
        super.init()
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    func sendLocation(_ locationMap: [String: Any?]) {
        eventSink?(locationMap)
    }
}

/// Stream handler for geofence events sent to Flutter.
class GeofenceStreamHandler: NSObject, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?

    override init() {
        eventSink = nil
        super.init()
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    func sendGeofenceEvent(_ eventMap: [String: Any?]) {
        eventSink?(eventMap)
    }
}

File: android/src/main/kotlin/io/github/bijoyknath/live_location_tracker_plus/LiveLocationTrackerPlusPlugin.kt
package io.github.bijoyknath.live_location_tracker_plus

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Build
import android.provider.Settings
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry

/** LiveLocationTrackerPlusPlugin */
class LiveLocationTrackerPlusPlugin :
    FlutterPlugin,
    MethodCallHandler,
    ActivityAware,
    PluginRegistry.RequestPermissionsResultListener {

    private lateinit var channel: MethodChannel
    private lateinit var locationEventChannel: EventChannel
    private lateinit var geofenceEventChannel: EventChannel
    private lateinit var context: Context

    private var activity: Activity? = null
    private var pendingPermissionResult: Result? = null
    private var pendingTrackingCall: MethodCall? = null
    private var pendingTrackingResult: Result? = null

    private var locationStreamHandler: LocationStreamHandler? = null
    private var geofenceStreamHandler: GeofenceStreamHandler? = null

    private var permissionHandler: PermissionHandler? = null
    private var geofenceManager: GeofenceManager? = null
    private var firebaseSyncManager: FirebaseSyncManager? = null

    private var isTrackingActive = false

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext

        // Method channel
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "live_location_tracker_plus")
        channel.setMethodCallHandler(this)

        // Location event channel
        locationStreamHandler = LocationStreamHandler()
        locationEventChannel = EventChannel(
            flutterPluginBinding.binaryMessenger,
            "live_location_tracker_plus/location_stream"
        )
        locationEventChannel.setStreamHandler(locationStreamHandler)

        // Geofence event channel
        geofenceStreamHandler = GeofenceStreamHandler()
        geofenceEventChannel = EventChannel(
            flutterPluginBinding.binaryMessenger,
            "live_location_tracker_plus/geofence_stream"
        )
        geofenceEventChannel.setStreamHandler(geofenceStreamHandler)

        // Managers
        permissionHandler = PermissionHandler(context)
        geofenceManager = GeofenceManager(context, geofenceStreamHandler)
        firebaseSyncManager = FirebaseSyncManager()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${Build.VERSION.RELEASE}")
            }
            "startTracking" -> handleStartTracking(call, result)
            "stopTracking" -> handleStopTracking(result)
            "getCurrentLocation" -> handleGetCurrentLocation(call, result)
            "isTracking" -> result.success(isTrackingActive)
            "addGeofence" -> handleAddGeofence(call, result)
            "removeGeofence" -> handleRemoveGeofence(call, result)
            "getActiveGeofences" -> handleGetActiveGeofences(result)
            "requestPermission" -> handleRequestPermission(result)
            "checkPermission" -> handleCheckPermission(result)
            "requestBackgroundPermission" -> handleRequestBackgroundPermission(result)
            "enableFirebaseSync" -> handleEnableFirebaseSync(call, result)
            "disableFirebaseSync" -> handleDisableFirebaseSync(result)
            "setTrackingMode" -> handleSetTrackingMode(call, result)
            "openLocationSettings" -> handleOpenLocationSettings(result)
            "openAppSettings" -> handleOpenAppSettings(result)
            else -> result.notImplemented()
        }
    }

    // -------------------------------------------------------------------------
    // Tracking
    // -------------------------------------------------------------------------

    private fun handleStartTracking(call: MethodCall, result: Result) {
        // Request notification permission on Android 13+ before starting foreground service
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            val act = activity
            if (act != null && androidx.core.content.ContextCompat.checkSelfPermission(
                    act, android.Manifest.permission.POST_NOTIFICATIONS
                ) != android.content.pm.PackageManager.PERMISSION_GRANTED
            ) {
                // Save pending call and wait for permission result
                pendingTrackingCall = call
                pendingTrackingResult = result
                androidx.core.app.ActivityCompat.requestPermissions(
                    act,
                    arrayOf(android.Manifest.permission.POST_NOTIFICATIONS),
                    1003
                )
                return
            }
        }

        doStartTracking(call, result)
    }

    private fun doStartTracking(call: MethodCall, result: Result) {
        val intervalMs = call.argument<Int>("intervalMs") ?: 5000
        val distanceFilter = call.argument<Double>("distanceFilter") ?: 10.0
        val accuracy = call.argument<Int>("accuracy") ?: 2
        val trackingMode = call.argument<Int>("trackingMode") ?: 1
        val showNotification = call.argument<Boolean>("showNotification") ?: true
        val notificationTitle = call.argument<String>("notificationTitle")
            ?: "Location Tracking Active"
        val notificationBody = call.argument<String>("notificationBody")
            ?: "Your location is being tracked in the background"
        val notificationChannelId = call.argument<String>("notificationChannelId")
            ?: "live_location_tracker_channel"
        val notificationIconName = call.argument<String>("notificationIconName")

        val serviceIntent = Intent(context, LocationService::class.java).apply {
            action = LocationService.ACTION_START
            putExtra(LocationService.EXTRA_INTERVAL_MS, intervalMs.toLong())
            putExtra(LocationService.EXTRA_DISTANCE_FILTER, distanceFilter.toFloat())
            putExtra(LocationService.EXTRA_ACCURACY, accuracy)
            putExtra(LocationService.EXTRA_TRACKING_MODE, trackingMode)
            putExtra(LocationService.EXTRA_SHOW_NOTIFICATION, showNotification)
            putExtra(LocationService.EXTRA_NOTIFICATION_TITLE, notificationTitle)
            putExtra(LocationService.EXTRA_NOTIFICATION_BODY, notificationBody)
            putExtra(LocationService.EXTRA_NOTIFICATION_CHANNEL_ID, notificationChannelId)
            putExtra(LocationService.EXTRA_NOTIFICATION_ICON_NAME, notificationIconName)
        }

        LocationService.setLocationStreamHandler(locationStreamHandler)
        LocationService.setFirebaseSyncManager(firebaseSyncManager)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            context.startForegroundService(serviceIntent)
        } else {
            context.startService(serviceIntent)
        }

        isTrackingActive = true
        result.success(true)
    }

    private fun handleStopTracking(result: Result) {
        val serviceIntent = Intent(context, LocationService::class.java).apply {
            action = LocationService.ACTION_STOP
        }
        context.startService(serviceIntent)
        isTrackingActive = false
        result.success(true)
    }

    private fun handleGetCurrentLocation(call: MethodCall, result: Result) {
        val accuracy = call.argument<Int>("accuracy") ?: 2
        LocationService.getCurrentLocation(context, accuracy) { locationMap ->
            if (locationMap != null) {
                result.success(locationMap)
            } else {
                result.error("LOCATION_ERROR", "Failed to get current location", null)
            }
        }
    }

    private fun handleSetTrackingMode(call: MethodCall, result: Result) {
        val mode = call.argument<Int>("mode") ?: 1
        val serviceIntent = Intent(context, LocationService::class.java).apply {
            action = LocationService.ACTION_UPDATE_MODE
            putExtra(LocationService.EXTRA_TRACKING_MODE, mode)
        }
        context.startService(serviceIntent)
        result.success(true)
    }

    // -------------------------------------------------------------------------
    // Geofencing
    // -------------------------------------------------------------------------

    private fun handleAddGeofence(call: MethodCall, result: Result) {
        val id = call.argument<String>("id") ?: return result.error("INVALID", "id required", null)
        val lat = call.argument<Double>("latitude")
            ?: return result.error("INVALID", "latitude required", null)
        val lng = call.argument<Double>("longitude")
            ?: return result.error("INVALID", "longitude required", null)
        val radius = call.argument<Double>("radius")?.toFloat()
            ?: return result.error("INVALID", "radius required", null)
        val triggers = call.argument<List<Int>>("triggers") ?: listOf(0, 1)
        val loiteringDelayMs = call.argument<Int>("loiteringDelayMs") ?: 0
        val expirationDurationMs = call.argument<Int>("expirationDurationMs")

        geofenceManager?.addGeofence(
            id, lat, lng, radius, triggers, loiteringDelayMs,
            expirationDurationMs?.toLong()
        ) { success ->
            result.success(success)
        }
    }

    private fun handleRemoveGeofence(call: MethodCall, result: Result) {
        val id = call.argument<String>("id")
            ?: return result.error("INVALID", "id required", null)
        geofenceManager?.removeGeofence(id) { success ->
            result.success(success)
        }
    }

    private fun handleGetActiveGeofences(result: Result) {
        val geofences = geofenceManager?.getActiveGeofences() ?: emptyList()
        result.success(geofences)
    }

    // -------------------------------------------------------------------------
    // Permissions
    // -------------------------------------------------------------------------

    private fun handleRequestPermission(result: Result) {
        val act = activity
        if (act == null) {
            result.error("NO_ACTIVITY", "Activity not available", null)
            return
        }
        pendingPermissionResult = result
        permissionHandler?.requestForegroundPermission(act)
    }

    private fun handleCheckPermission(result: Result) {
        val status = permissionHandler?.checkPermission() ?: 0
        result.success(status)
    }

    private fun handleRequestBackgroundPermission(result: Result) {
        val act = activity
        if (act == null) {
            result.error("NO_ACTIVITY", "Activity not available", null)
            return
        }
        pendingPermissionResult = result
        permissionHandler?.requestBackgroundPermission(act)
    }

    // -------------------------------------------------------------------------
    // Firebase Sync
    // -------------------------------------------------------------------------

    private fun handleEnableFirebaseSync(call: MethodCall, result: Result) {
        val collectionPath = call.argument<String>("collectionPath") ?: "live_locations"
        val userId = call.argument<String>("userId")
            ?: return result.error("INVALID", "userId required", null)
        val syncIntervalMs = call.argument<Int>("syncIntervalMs") ?: 10000

        val success = firebaseSyncManager?.enable(collectionPath, userId, syncIntervalMs.toLong())
            ?: false
        result.success(success)
    }

    private fun handleDisableFirebaseSync(result: Result) {
        firebaseSyncManager?.disable()
        result.success(true)
    }

    // -------------------------------------------------------------------------
    // Settings
    // -------------------------------------------------------------------------

    private fun handleOpenLocationSettings(result: Result) {
        try {
            val intent = Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS).apply {
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startActivity(intent)
            result.success(true)
        } catch (e: Exception) {
            result.success(false)
        }
    }

    private fun handleOpenAppSettings(result: Result) {
        try {
            val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                data = android.net.Uri.parse("package:${context.packageName}")
            }
            context.startActivity(intent)
            result.success(true)
        } catch (e: Exception) {
            result.success(false)
        }
    }

    // -------------------------------------------------------------------------
    // Permission Result Callback
    // -------------------------------------------------------------------------

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ): Boolean {
        if (requestCode == 1003) {
            // Notification permission result — resume pending tracking
            val call = pendingTrackingCall
            val res = pendingTrackingResult
            pendingTrackingCall = null
            pendingTrackingResult = null
            if (call != null && res != null) {
                doStartTracking(call, res)
            }
            return true
        }

        val status = permissionHandler?.handlePermissionResult(requestCode, permissions, grantResults) ?: 0
        pendingPermissionResult?.success(status)
        pendingPermissionResult = null
        return true
    }

    // -------------------------------------------------------------------------
    // ActivityAware
    // -------------------------------------------------------------------------

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addRequestPermissionsResultListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addRequestPermissionsResultListener(this)
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    // -------------------------------------------------------------------------
    // Cleanup
    // -------------------------------------------------------------------------

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        // Null out sinks so service doesn't try to send to dead Flutter engine
        locationStreamHandler?.clearSink()
        geofenceStreamHandler?.clearSink()
        locationEventChannel.setStreamHandler(null)
        geofenceEventChannel.setStreamHandler(null)
    }
}

// =============================================================================
// Stream Handlers
// =============================================================================

/** Stream handler for location updates. */
class LocationStreamHandler : EventChannel.StreamHandler {
    private var eventSink: EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    fun sendLocation(locationMap: Map<String, Any?>) {
        try {
            eventSink?.success(locationMap)
        } catch (e: Exception) {
            // Flutter engine detached — service keeps running silently
        }
    }

    fun clearSink() {
        eventSink = null
    }
}

/** Stream handler for geofence events. */
class GeofenceStreamHandler : EventChannel.StreamHandler {
    private var eventSink: EventChannel.EventSink? = null

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    fun sendGeofenceEvent(eventMap: Map<String, Any?>) {
        try {
            eventSink?.success(eventMap)
        } catch (e: Exception) {
            // Flutter engine detached — service keeps running silently
        }
    }

    fun clearSink() {
        eventSink = null
    }
}
